timeline/src/Data/Timeline/Internal.hs

```diff
@@ -0,0 +1,299 @@
+
+data TimeRange = TimeRange
+  { trFrom :: Maybe UTCTime,
+    trTo :: Maybe UTCTime
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+
+isTimeAfterRange :: UTCTime -> TimeRange -> Bool
+isTimeAfterRange t TimeRange {trTo} = maybe False (t >=) trTo
+
+instance FunctorWithIndex TimeRange Timeline where
+  imap :: (TimeRange -> a -> b) -> Timeline a -> Timeline b
+  imap f Timeline {..} =
+    Timeline
+      { initialValue = f initialRange initialValue,
+        values = flip Map.mapWithKey values $ \from value ->
+          let timeRange = TimeRange (Just from) (fst <$> Map.lookupGT from values)
+           in f timeRange value
+      }
+    where
+      initialRange = TimeRange Nothing $ fst <$> Map.lookupMin values
+
+instance FoldableWithIndex TimeRange Timeline
+
+instance TraversableWithIndex TimeRange Timeline where
+  itraverse :: (Applicative f) => (TimeRange -> a -> f b) -> Timeline a -> f (Timeline b)
+  itraverse f = sequenceA . imap f
+
+changes :: Timeline a -> Set UTCTime
+changes Timeline {values} = Map.keysSet values
+
+data Record a = Record
+  { effectiveFrom :: UTCTime,
+    effectiveTo :: Maybe UTCTime,
+    value :: a
+  }
+  deriving stock (Eq, Functor)
+
+type Getter s a = forall f. (Contravariant f) => (a -> f a) -> s -> f s
+
+recordEffectiveFrom :: Getter (Record a) UTCTime
+recordEffectiveFrom f r = contramap getEffectiveFrom (f (getEffectiveFrom r))
+  where
+    getEffectiveFrom :: Record a -> UTCTime
+    getEffectiveFrom Record {effectiveFrom} = effectiveFrom
+
+recordEffectiveTo :: Getter (Record a) (Maybe UTCTime)
+recordEffectiveTo f r = contramap getEffectiveTo (f (getEffectiveTo r))
+  where
+    getEffectiveTo :: Record a -> Maybe UTCTime
+    getEffectiveTo Record {effectiveTo} = effectiveTo
+
+recordValue :: Getter (Record a) a
+recordValue f r = contramap getValue (f (getValue r))
+  where
+    getValue :: Record a -> a
+    getValue Record {value} = value
+
+-- | 'makeRecord' returns 'Nothing' if @effectiveTo@ is not greater than @effectiveFrom@
+makeRecord ::
+  -- | effective from
+  UTCTime ->
+  -- | optional effective to
+  Maybe UTCTime ->
+  -- | value
+  a ->
+  Maybe (Record a)
+-- can't write a makeRecordTH because UTCTime has no Lift instance
+makeRecord effectiveFrom effectiveTo value =
+  if maybe False (effectiveFrom >=) effectiveTo
+    then Nothing
+    else Just Record {..}
+
+instance (TH.Lift a) => TH.Lift (Record a) where
+  liftTyped Record {..} =
+    [||
+    Record
+      (unLiftUTCTime $$(TH.liftTyped $ LiftUTCTime effectiveFrom))
+      (fmap unLiftUTCTime $$(TH.liftTyped $ LiftUTCTime <$> effectiveTo))
+      $$(TH.liftTyped value)
+    ||]
+
+newtype LiftUTCTime = LiftUTCTime UTCTime
+  deriving stock (Generic)
+
+unLiftUTCTime :: LiftUTCTime -> UTCTime
+unLiftUTCTime (LiftUTCTime t) = t
+
+instance TH.Lift LiftUTCTime where
+  liftTyped (LiftUTCTime t) = [||LiftUTCTime (read $$(TH.liftTyped (show t)))||]
```

Comment:
I would probably go via a two-tuple instead of relying on `read`/`show` to magically do the right thing.

Reply:
I switched from our existing solution because `YearDay` doesn't compile under GHC 8.10.7. Yes I agree not using `read`/`show` is safer.